src/ripple/app/reporting/ETLSource.h

```diff
@@ -0,0 +1,435 @@
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CORE_ETLSOURCE_H_INCLUDED
+#define RIPPLE_CORE_ETLSOURCE_H_INCLUDED
+#include <ripple/app/main/Application.h>
+#include <ripple/app/reporting/ETLHelpers.h>
+#include <ripple/protocol/STLedgerEntry.h>
+#include <ripple/rpc/Context.h>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/beast/core.hpp>
+#include <boost/beast/core/string.hpp>
+#include <boost/beast/websocket.hpp>
+
+#include "org/xrpl/rpc/v1/xrp_ledger.grpc.pb.h"
+#include <grpcpp/grpcpp.h>
+
+namespace ripple {
+
+class ReportingETL;
+
+/// This class manages a connection to a single ETL source. This is almost
+/// always a p2p node, but really could be another reporting node. This class
+/// subscribes to the ledgers and transactions_proposed streams of the
+/// associated p2p node, and keeps track of which ledgers the p2p node has. This
+/// class also has methods for extracting said ledgers. Lastly this class
+/// forwards transactions received on the transactions_proposed streams to any
+/// subscribers.
+class ETLSource
+{
+    std::string ip_;
+
+    std::string wsPort_;
+
+    std::string grpcPort_;
+
+    ReportingETL& etl_;
+
+    // a reference to the applications io_service
+    boost::asio::io_context& ioc_;
+
+    std::unique_ptr<org::xrpl::rpc::v1::XRPLedgerAPIService::Stub> stub_;
+
+    std::unique_ptr<boost::beast::websocket::stream<boost::beast::tcp_stream>>
+        ws_;
+    boost::asio::ip::tcp::resolver resolver_;
+
+    boost::beast::flat_buffer readBuffer_;
+
+    std::vector<std::pair<uint32_t, uint32_t>> validatedLedgers_;
+
+    std::string validatedLedgersRaw_;
+
+    NetworkValidatedLedgers& networkValidatedLedgers_;
+
+    beast::Journal journal_;
+
+    Application& app_;
+
+    std::mutex mtx_;
+
+    size_t numFailures_ = 0;
+
+    std::atomic_bool closing_ = false;
+
+    std::atomic_bool connected_ = false;
+
+    // true if this ETL source is forwarding transactions received on the
+    // transactions_proposed stream. There are usually multiple ETL sources,
+    // so to avoid forwarding the same transaction multiple times, we only
+    // forward from one particular ETL source at a time.
+    std::atomic_bool forwardingStream_ = false;
+
+    // The last time a message was received on the ledgers stream
+    std::chrono::time_point<std::chrono::system_clock> lastMsgTime_;
```

Comment:
You can get rid of one of the `std::chrono` by doing:
```c++
std::chrono::system_clock::time_point lastMsgTime_;
```
It is the exact same type either way.

Reply:
changed